### ISSUE TITLE
XLA Compilation fails: Non-root tuple types are not handled

### ISSUE THREAD

**Issue by janbernloehr** [tensorflow/tensorflow#53123]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 20.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: No
- TensorFlow installed from (source or binary): Yes
- TensorFlow version (use command below): 2.7.0
- Python version: 3.8.0
- Bazel version (if compiling from source): -
- GCC/Compiler version (if compiling from source): -
- CUDA/cuDNN version: 11.4.2 / 8.2.4
- GPU model and memory: RTX 2080 Ti 10GB

You can collect some of this information using our environment capture
```
python version: 3.8.10
python branch: 
python build version: ('default', 'Sep 28 2021 16:10:42')
python compiler version: GCC 9.3.0
python implementation: CPython

tf.version.VERSION = 2.7.0
tf.version.GIT_VERSION = v2.7.0-0-gc256c071
tf.version.COMPILER_VERSION = 9.3.0
```

**Describe the current behavior**
When running our custom training loop with `TF_XLA_FLAGS=--tf_xla_auto_jit=2`, the following error is obtained

```
2021-11-19 01:57:34.543515: W tensorflow/core/framework/op_kernel.cc:1745] OP_REQUIRES failed at xla_ops.cc:432 : INTERNAL: Non-root tuple types are not handled.
Traceback (most recent call last):
  File "train.py", line 140, in <module>
    main(sys.argv[1:])
  File "train.py", line 130, in main
    trainer.train(
  File "/home/bje2lr/venv/cu114/lib/python3.8/site-packages/tensorflow/python/util/traceback_utils.py", line 153, in error_handler
    raise e.with_traceback(filtered_tb) from None
  File "/home/bje2lr/venv/cu114/lib/python3.8/site-packages/tensorflow/python/eager/execute.py", line 58, in quick_execute
    tensors = pywrap_tfe.TFE_Py_Execute(ctx._handle, device_name, op_name,
tensorflow.python.framework.errors_impl.InternalError:  Non-root tuple types are not handled.
         [[{{node cluster_3_1/xla_compile}}]] [Op:__inference_distributed_train_step_56538]
```

**Describe the expected behavior**

`tf_xla_auto_jit=2` should only jit functions that are supported by XLA.

**[Contributing](https://www.tensorflow.org/community/contribute)**

- Do you want to contribute a PR? (yes/no): no
- Briefly describe your candidate solution(if contributing): -

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

Unfortunatelly, uploading the xla log files did not work.

**Comment by janbernloehr** [comment]
Ok, the issue was probably fixed in 1d77ab0bbf030462d15afd2de435676c46b04c7c or a related commit.

**Comment by mohantym** [comment]
Hi @janbernloehr ! Could you attach sample standalone code to reproduce this issue too? Thanks!

**Comment by leondgarse** [comment]
This is what I dig out from my project. Just a standalone code for reproducing, but it maybe my own case.
- This one will throw error `Non-root tuple types are not handled` in `TF 2.7.0` with `TF_XLA_FLAGS="--tf_xla_auto_jit=2`
  ```py
  tf.__version__
  # 2.7.0
  xx = tf.random.uniform((1000, 32, 32, 3))
  yy = tf.one_hot(tf.cast(tf.random.uniform((1000,)) * 10, 'int32'), depth=32)
  mm = keras.models.Sequential([keras.layers.Input([32, 32, 3]), keras.layers.Flatten(), keras.layers.Dense(32)])
  mm.compile(optimizer='SGD', loss="categorical_crossentropy", metrics='acc')
  mm.fit(xx, yy)
  # InternalError:  Non-root tuple types are not handled.
  #          [[{{node cluster_0_1/xla_compile}}]] [Op:__inference_train_function_498]
  ```
- Just use output Dense `31` instead of `32` will work normally
  ```py
  tf.__version__
  # 2.7.0
  xx = tf.random.uniform((1000, 32, 32, 3))
  yy = tf.one_hot(tf.cast(tf.random.uniform((1000,)) * 10, 'int32'), depth=31)
  mm = keras.models.Sequential([keras.layers.Input([32, 32, 3]), keras.layers.Flatten(), keras.layers.Dense(31)])
  mm.compile(optimizer='SGD', loss="categorical_crossentropy", metrics='acc')
  mm.fit(xx, yy)
  ```
- Use output Dense `32` without `metrics='acc'` also works
  ```py
  tf.__version__
  # 2.7.0
  xx = tf.random.uniform((1000, 32, 32, 3))
  yy = tf.one_hot(tf.cast(tf.random.uniform((1000,)) * 10, 'int32'), depth=32)
  mm = keras.models.Sequential([keras.layers.Input([32, 32, 3]), keras.layers.Flatten(), keras.layers.Dense(32)])
  mm.compile(optimizer='SGD', loss="categorical_crossentropy")
  mm.fit(xx, yy)
  ```
Yes, theses test cases all pass using latest `tf-nightly 2.8.0-dev20211201` and `TF 2.6.2`. I think maybe it's really fixed.

**Comment by aliencaocao** [comment]
> Ok, the issue was probably fixed in [1d77ab0](https://github.com/tensorflow/tensorflow/commit/1d77ab0bbf030462d15afd2de435676c46b04c7c) or a related commit.

I run into the same issue today on TF2.7 and after upgrading to 2.8 it is indeed fixed.
